### PR TITLE
Getting Glow compiling and running under Win32.

### PR DESCRIPTION
**Pull request by lygstate**
Getting OpenCL works under Win32.
CPU backend still not works and disabled now for Win32.

### CODE REVIEW

**Comment by lygstate** [comment]
What's wrong with the CLA? Why I have the following issue:
```
There were some errors submitting your CLA:
GitHub Username is already associated with an existing contributor.
Email is already associated with an existing contributor.
If you would like to update an existing contributor account, please email cla@fb.com with your details.
```

**Comment by opti-mix** [comment]
@compnerd You may want to have a look at this.

**Comment by nadavrot** [comment]
@lygstate Thanks for doing this work!  I am not sure what's wrong with the CLA.

File: lib/Graph/Graph.cpp [hunk 1]
```diff
@@ -581,7 +581,7 @@ Node *Function::createBroadcast(llvm::StringRef name, NodeValue input,
   // new shape (no action taken) or == 1 (broadcast in that direction). Else
   // the original shape had no dimensions here (after considering axis), so
   // add the new dimension and broadcast in that direction.
-  size_t reshapeDims[newShape.size()];
```
Comment (nadavrot) [inline lib/Graph/Graph.cpp]: GLOW_VLA is not very readable. I prefer to just use array of size "max_tensor_dimensions" here.

File: lib/Base/Tensor.cpp [hunk 2]
```diff
@@ -219,8 +220,8 @@ template <class ElemTy>
 static bool tryTransposeFastImpl(Handle<ElemTy> &src, Handle<ElemTy> &dest,
                                  llvm::ArrayRef<unsigned> shuffle) {
   const size_t numDims = dest.dims().size();
-  size_t srcCoorArr[numDims];
```
Comment (nadavrot) [inline lib/Base/Tensor.cpp]: GLOW_VLA is not very readable. I prefer to just use array of size "max_tensor_dimensions" here.

File: lib/Backends/OpenCL/kernels.cl [hunk 3]
```diff
@@ -6,20 +6,26 @@ typedef unsigned cl_uint32_t;
 /// This type is always 64 bits.
 typedef unsigned long cl_uint64_t;
 
+#if GLOW_HOST_SIZEOF_SIZE_T == 4
```
Comment (nadavrot) [inline lib/Backends/OpenCL/kernels.cl]: This is not correct. The host does not have to match the device.

File: lib/Backends/OpenCL/OpenCL.cpp [hunk 4]
```diff
@@ -692,8 +713,9 @@ void OCLBackend::doForwardPass() {
           createKernel(useTiledMatMul ? "matmul_tiled" : kernelName);
       setKernelArg(kernel, 0, deviceBuffer_);
 
-      unsigned numArgs = I.getNumOperands();
-      for (unsigned arg = 0; arg < numArgs; arg++) {
+      /* unsigned numArgs = I.getNumOperands(); */
```
Comment (nadavrot) [inline lib/Backends/OpenCL/OpenCL.cpp]: What is this?

File: include/glow/Support/Support.h [hunk 5]
```diff
@@ -23,6 +23,21 @@
 #include <iostream>
 #include <sstream>
 
+#ifdef _MSC_VER
```
Comment (nadavrot) [inline include/glow/Support/Support.h]: I prefer to work around the dynamic use and not declare GLOW_VLA which is not very readable.

File: cmake/modules/FindOpenCL.cmake [hunk 6]
```diff
@@ -0,0 +1,184 @@
+# Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
```
Comment (nadavrot) [inline cmake/modules/FindOpenCL.cmake]: Where did you get this file from? Isn't this the same file as the one that comes with CMake? Why do we need it here?

File: CMakeLists.txt [hunk 7]
```diff
@@ -86,5 +99,5 @@ add_custom_target(dependency_graph
 # Fetch the dependencies for all the tests.
 get_property(GLOW_TEST_DEPENDS GLOBAL PROPERTY GLOW_TEST_DEPENDS)
 
-add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND}
+add_custom_target(check-glow COMMAND ${CMAKE_CTEST_COMMAND}
```
Comment (nadavrot) [inline CMakeLists.txt]: Why did you change this?

File: tests/unittests/CMakeLists.txt [hunk 8]
```diff
@@ -1,3 +1,11 @@
+if (MSVC)
```
Comment (compnerd) [inline tests/unittests/CMakeLists.txt]: This seems like the result of not building googletest

File: tools/loader/CMakeLists.txt [hunk 9]
```diff
@@ -10,3 +10,7 @@ target_link_libraries(image-classifier
                         ExecutionEngine
                         IR
                         Quantization)
+
+target_include_directories(image-classifier
+  PRIVATE
```
Comment (compnerd) [inline tools/loader/CMakeLists.txt]: Please fix the indentation

File: tests/unittests/CMakeLists.txt (hunk 8 above)
Comment (lygstate) [reply tests/unittests/CMakeLists.txt]: llvm also have googletest and test target

File: cmake/modules/FindOpenCL.cmake (hunk 6 above)
Comment (lygstate) [reply cmake/modules/FindOpenCL.cmake]: The one in CMake doesn't works properly under win32， I have to copy and modify

File: lib/Graph/Graph.cpp (hunk 1 above)
Comment (lygstate) [reply lib/Graph/Graph.cpp]: VLA means Variable Length Array， msvc doesn't support for that， so I place it in support.h， the name comes form 
https://github.com/jerryscript-project/jerryscript/blob/a74bf7dc91e7358f0e6c7e111e727110377fca82/jerry-core/include/jerryscript-compiler.h#L62,
Also are my patch, I hope this would have an agreement

File: lib/Backends/OpenCL/OpenCL.cpp (hunk 4 above)
Comment (lygstate) [reply lib/Backends/OpenCL/OpenCL.cpp]: An bug, test case show it was 4, but we expected 3

Comment (nadavrot) [reply lib/Backends/OpenCL/OpenCL.cpp]: @opti-mix Do you mind looking at this bug?

File: lib/Graph/Graph.cpp (hunk 1 above)
Comment (nadavrot) [reply lib/Graph/Graph.cpp]: I know what VLA is. I am saying that GLOW_VLA is not very readable. I prefer to use the static size: int foo[max_tensor_dimensions];

Comment (lygstate) [reply lib/Graph/Graph.cpp]: so does max_tensor_dimensions a constants or a macro？

Comment (nadavrot) [reply lib/Graph/Graph.cpp]: It does not matter if it’s a macro or a constant number. It is not dynamic.

Comment (lygstate) [reply lib/Graph/Graph.cpp]: I'll try constant number first, if doesn't works, macro would used.

File: lib/Backends/OpenCL/OpenCL.cpp (hunk 4 above)
Comment (opti-mix) [reply lib/Backends/OpenCL/OpenCL.cpp]: @lygstate I checked under debugger and I see I.getNumOperands() == 3 for MatMulInst instructions. It is also defined with 3 arguments in InstrGen.cpp. 

If you have discovered a case, where it reported 4 arguments, could you provide a description how to reproduce this?

File: CMakeLists.txt [hunk 10]
```diff
@@ -54,26 +57,36 @@ if (GLOW_WITH_OPENCL)
...
-endif()
-if (NOT LLVM_FOUND)
-  message(FATAL_ERROR "Could not find LLVM. Build LLVM manually and configure the project with -DCMAKE_PREFIX_PATH=/path/to/llvm/install")
+if(MSVC)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4710 /wd4711 /wd4514 /wd4571 /wd4820 /wd4774 /wd4625")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4526 /wd4061 /wd4365 /wd4242 /wd4996 /wd5026 /wd4626")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd5027 /wd4668 /wd4305 /wd4826 /wd4574")
+else()
+  # Prefer LLVM 6.
+  find_package(LLVM 6 CONFIG)
```
Comment (bertmaher) [inline CMakeLists.txt]: Something doesn't make sense here... llvm is a required dependency of glow.  I don't understand how you can possibly be building without it.

File: lib/Backends/OpenCL/OpenCL.cpp (hunk 4 above)
Comment (lygstate) [reply lib/Backends/OpenCL/OpenCL.cpp]:
```

TEST_P(Operator, simplePredication) {
  auto *inputs = mod_.createVariable(ElemKind::FloatTy, {10, 10, 10}, "inputs");
  auto *counters = mod_.createVariable(ElemKind::FloatTy, {10}, "counters");

  counters->getHandle() = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
  inputs->getHandle().randomize(-10, 10, mod_.getPRNG());

  Node *C5 = F_->createSplat("C5", counters->getType(), 5.0);
  Node *pred = F_->createCmpLTE("cmp", C5, counters);

  auto *FC0 = F_->createFullyConnected("FC0", inputs, 128);
  auto *RL0 = F_->createRELU("RL0", FC0);
  auto *FC1 = F_->createFullyConnected("FC1", RL0, 64);
  auto *RL1 = F_->createRELU("RL1", FC1);
  auto *FC2 = F_->createFullyConnected("FC2", RL1, 32);
  auto *RL2 = F_->createRELU("RL2", FC2);

  F_->createSave("ret", RL2);

  FC0->setPredicate(pred);
  FC1->setPredicate(pred);
  FC2->setPredicate(pred);

  EE_.compile(CompilationMode::Infer, F_);
  EE_.run({}, {});
}
```
The OperatorTest.cpp
simplePredication

File: CMakeLists.txt (hunk 10 above)
Comment (lygstate) [reply CMakeLists.txt]: @bertmaher , Win32 have no llvm-devel package, I building glow with llvm source.
[

File: lib/Backends/OpenCL/OpenCL.cpp (hunk 4 above)
Comment (lygstate) [reply lib/Backends/OpenCL/OpenCL.cpp]: I have already use the CI to trigger the issue

File: lib/Backends/OpenCL/kernels.cl (hunk 3 above)
Comment (lygstate) [reply lib/Backends/OpenCL/kernels.cl]: For quick fix the inconsistence between Host and Device.

File: lib/Backends/CPU/libjit/libjit_defs.h [hunk 11]
```diff
@@ -26,7 +26,7 @@ typedef __declspec(align(64)) float float16;
 #else
 typedef float float4 __attribute__((ext_vector_type(4)));
 typedef float float8 __attribute__((ext_vector_type(8)));
-typedef float float16 __attribute__((ext_vector_type(16)));
+typedef float float16 __attribute__((aligned(64)));
```
Comment (nadavrot) [inline lib/Backends/CPU/libjit/libjit_defs.h]: This change is not correct. Float16 would be a regular float that's aligned to 64b.

**Comment by nadavrot** [comment]
@lygstate This PR has a ton of good stuff. Would it be possible to extract more changes out of this PR and start submitting them as small focused PRs?  I think that this work is awesome and it would be great to see these PRs land.

**Comment by lygstate** [comment]
@nadavrot I am waiting for feedback, which part I should split out as smaller patch:)
